hydra-node/src/Hydra/HeadLogic.hs

```diff
@@ -195,6 +195,118 @@ data Environment = Environment
+processL2OutCome (L2NewState (nextState, effects)) parameters previousRecoverableState =
+  let nextState' = processL2NextState nextState parameters previousRecoverableState
+   in NewState nextState' effects
+processL2OutCome (L2Wait reason) _ _ = Wait reason
+
+-- | On IdleState, upon client init tx
+--   1. build effects to produce
+--   2. stay in same IdleState
+onIdleHandleClientInitTxEvent :: HeadParameters -> L2Outcome tx
+onIdleHandleClientInitTxEvent parameters =
+  L2NewState (SameState, effects)
+ where
+  effects = [OnChainEffect (InitTx parameters)]
+
+-- | On IdleState, upon chain init tx
+--   1. build effects to produce
+--   2. move to new InitialState
+onIdleHandleChainInitTxEvent :: [Party] -> L2Outcome tx
+onIdleHandleChainInitTxEvent parties =
+  L2NewState (nextState, effects)
+ where
+  nextState = NextInitialState{pendingCommits = Set.fromList parties, committed = mempty}
+  effects = [ClientEffect $ ReadyToCommit $ fromList parties]
+
+-- | On InitialState, upon client commit tx
+--   1. build effects to produce
+--   2. stay in same InitialState
+onInitialHandleClientCommitTxEvent :: Party -> UTxOType tx -> L2Outcome tx
+onInitialHandleClientCommitTxEvent party utxo =
+  L2NewState (SameState, effects)
+ where
+  effects = [OnChainEffect (CommitTx party utxo)]
+
+-- | On InitialState, upon chain commit tx
+--   1. build effects to produce
+--   2. move to new InitialState
+onInitialHandleChainCommitTxEvent ::
+  Monoid (UTxOType tx) =>
+  Party ->
+  Party ->
+  UTxOType tx ->
+  Set Party ->
+  Map Party (UTxOType tx) ->
+  L2Outcome tx
+onInitialHandleChainCommitTxEvent pt party utxo pendingCommits committed =
+  L2NewState (nextState, effects)
+ where
+  nextState = NextInitialState{pendingCommits = remainingParties, committed = newCommitted}
+  effects =
+    [ClientEffect $ Committed pt utxo]
+      <> [OnChainEffect $ CollectComTx collectedUTxO | canCollectCom]
+  remainingParties = Set.delete pt pendingCommits
+  newCommitted = Map.insert pt utxo committed
+  canCollectCom = null remainingParties && pt == party
+  collectedUTxO = mconcat $ Map.elems newCommitted
+
+-- | On InitialState, upon chain collect tx
+--   1. build coordinatedHeadState
+--   2. build effects to produce
+--   3. move to new OpenState
+onIdleHandleChainCollectTxEvent ::
+  (Foldable t, Monoid (UTxOType tx)) => t (UTxOType tx) -> L2Outcome tx
+onIdleHandleChainCollectTxEvent committed =
+  let u0 = fold committed
+      effects = [ClientEffect $ HeadIsOpen u0]
+      coordinatedHeadState =
+        CoordinatedHeadState u0 mempty (InitialSnapshot $ Snapshot 0 u0 mempty) NoSeenSnapshot
+   in L2NewState (NextOpenState{coordinatedHeadState}, effects)
+
+-- | On OpenState, upon network request tx
+--   1. evaluate eitherTx
+--    a/ if the transaction is not applicable yet, wait on not applicable tx and produce no effects
+--    b/ otherwise, move to new OpenState using appied tx and produce client tx seen effects
```

Comment:
Yeah the `evaluate eitherTx` is what actually matters here and can't stay outside of this function. In the paper it's written like:
![image](https://user-images.githubusercontent.com/2621189/181182279-b5d69827-6dd2-4db2-89e6-63b2fafef9d2.png)
where the 
![image](https://user-images.githubusercontent.com/2621189/181182340-bfb4ac4b-e590-4aa6-8d48-cb078cb9b975.png)
is exactly the juicy bit you left outside (the `eitherTx`)